Gdt 100 auth lambda

JPrevost opened the pull request:
Why are these changes being introduced:

* This adds the logic to check for JWT tokens and redirect if the token
  is missing, invalid, or expired
* This follows the build process expected by our infrastructure code

Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/GDT-100

### Includes new or updated dependencies?

YES

### Developer

- [x] The `Dev CF Lambda@Edge Full Deploy` workflow has been run to update CloudFront in Dev1
- [x] All related Jira tickets are linked in commit message(s)
- [ ] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)

- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no change

JPrevost committed:
Adds initial auth redirect lambda and built process

Why are these changes being introduced:

* This adds the logic to check for JWT tokens and redirect if the token
  is missing, invalid, or expired
* This follows the build process expected by our infrastructure code

Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/GDT-100

ghukill commented on lambdas/lambda_edge.py:
```diff
@@ -1,10 +1,107 @@
+        Path="/apps/cf-lambda-geo-auth/", WithDecryption=False
+    )
+
+    logging.info(ssm_params)
```
Do we have concerns with logging SSM parameters?  Looks as though one parameter is the JWT secret which could be sensitive.

JPrevost replied:
This was a leftover debug log I had dropped in and forgot to remove. Technically anyone that can read the logs can probably also read the SSM param in the console so it's only a pseudo-secret but I do agree it'll be better to remove this line.

ghukill commented on lambdas/lambda_edge.py:
```diff
@@ -1,10 +1,107 @@
+def parse_cookies(headers):
+    parsed_cookie = {}
+    if headers.get("cookie"):
+        for cookie in headers["cookie"][0]["value"].split(";"):
```
The [Lambda@Edge docs](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-event-structure.html#lambda-event-structure-request) don't seem to have an example, but is it possible there could be multiple cookies in this `headers["cookie"]` array?  If so, are we safe just grabbing the 0th one?

The AWS "Q" chat thing suggests they could look like:
```json
[
  {
    "key": "cookie1", 
    "value": "cookie1Value"
  },
  {
    "key": "cookie2",
    "value": "cookie2Value" 
  }
]
```

And/or, do we know the name of the cookie and could use the `key` property to find it?

JPrevost replied:
Great question. I'll see if I can find more info on real-world scenarios on what to expect here.

JPrevost replied:
I've opened a separate ticket to investigate and update code if appropriate for the cookie array

https://mitlibraries.atlassian.net/browse/GDT-240

ghukill commented on lambdas/lambda_edge.py:
```diff
@@ -1,10 +1,107 @@
+            jwt_secret = param["Value"]
+        if param["Name"] == "/apps/cf-lambda-geo-auth/auth-url":
+            auth_url = param["Value"]
+
```
Wonder if we should throw an exception here if not all the expected SSM params are found?  

The advantage I see would be easy debugging if the lambda and TF code got out of sync, where otherwise we'd just get errors about undefined variables.

ghukill commented on tests/test_lambda_edge.py:
```diff
@@ -1,5 +1,354 @@
+@pytest.fixture
+def headers_with_cookies(valid_token):
+    return {
+        "host": [{"key": "Host", "value": "d123.cf.net"}],
+        "cookie": [
+            {
+                "key": "Cookie",
+                "value": f"mitlcdnauthjwt={valid_token}; AnotherOne=A; X-Name=B",
+            }
+        ],
```
Maybe this fixture answers my question about the 0th cookie above... looks like all the `header` dictionary values are lists?  If so, makes sense that it could be safe to the take the 0th one.

JPrevost replied:
Yeah it's definitely odd as it's an array but it feels like it doesn't function as one. I'll dig in more on this as it should be fairly easy for me to set a bunch of cookies locally and see how they get treated.

JPrevost replied:
I've opened a separate ticket to investigate and update code if appropriate for the cookie array

https://mitlibraries.atlassian.net/browse/GDT-240

ghukill reviewed (commented):
Overall, looks good to me!

Feel like the organization of code and functionality matches the relative simplicity of the lambda itself (e.g. pondered if a standalone function might be helpful for constructing the redirect URL, but didn't feel necessary).

Before approval, left a couple of questions, with the two that gave me approval pause being:
- potential logging of SSM params
- consider throwing explicit exception if SSM params not set

JPrevost committed:
Extracts ssm lookups to a separate method

Why are these changes being introduced:

* Code review suggested we throw exceptions if the parameters were not
  found in SSM. This led to more complexity in the code block which
  felt like it was easier to understand when extracted to it's own
  method.

Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/GDT-100

Document any side effects to this change:

* Some method documentation has been updated to better describe the core
  logic of the lambda

ghukill commented on lambdas/lambda_edge.py:
```diff
@@ -52,40 +52,62 @@ def decode_jwt(usertoken, jwt_secret):
+    """
+    request = event["Records"][0]["cf"]["request"]
+    headers = request["headers"]
+    jwt_secret, auth_url = load_ssm_params()
```
Think that was a good call to break out into another function.

ghukill commented on lambdas/lambda_edge.py:
```diff
@@ -52,40 +52,62 @@ def decode_jwt(usertoken, jwt_secret):
 
-    # Check for session-id in cookie, if present, then proceed with request
+    # confirm expected SSM params loaded properly
+    if jwt_secret is None:
+        msg = "/apps/cf-lambda-geo-auth/jwt-secret not found in parameter store"
+        logging.exception(msg)
+        raise RuntimeError(msg)
+
+    if auth_url is None:
+        msg = "/apps/cf-lambda-geo-auth/auth-url not found in parameter store"
+        logging.exception(msg)
+        raise RuntimeError(msg)
```
I dig it!  Thanks for adding.

ghukill reviewed (approved):
Looks good to me!  The updates addressed any concerns, and makes sense about a ticket to look into cookie structure later.
